### PR TITLE
Added retry logic & batching settings

### PR DESCRIPTION
**Pull request by shakeelmohamed**

### CODE REVIEW

File: splunklogger.js [hunk 1]
```diff
@@ -107,6 +136,58 @@ var defaultRequestOptions = {
...
+            throw new Error("Batch interval must be a positive number, found: " + interval);
+        }
+        else {
+            if (this._timerID) {
+                this._disableTimer();
+            }
+            
+            // If batch interval is changed, update the config property
+            if (this.config) {
+                this.config.batchInterval = interval;
```
Comment (itay) [inline splunklogger.js]: Why do you need to do this?

File: splunklogger.js [hunk 2]
```diff
@@ -107,6 +136,58 @@ var defaultRequestOptions = {
...
+            }
+
+            this._timerDuration = interval;
+
+            var that = this;
+            this._timerID = setInterval(function() {
```
Comment (itay) [inline splunklogger.js]: It's not a big deal, but it seems like you could have a private method on the object, `_timerFlush()`, which would do this logic. That way you don't need to recreate a new one every time.

File: splunklogger.js [hunk 3]
```diff
@@ -169,16 +250,66 @@ SplunkLogger.prototype._initializeConfig = function(config) {
         ret.protocol = config.protocol || ret.protocol || defaultConfig.protocol;
         ret.level = config.level || ret.level || defaultConfig.level;
 
+        ret.maxRetries = config.maxRetries || ret.maxRetries || defaultConfig.maxRetries;
```
Comment (itay) [inline splunklogger.js]: You repeat this pattern a few times - it may be good to put them in a utility function.

File: splunklogger.js (hunk 1 above)
Comment (shakeelmohamed) [inline splunklogger.js]: If `_enableTimer(interval)` is called explicitly, we just store that interval value in the config. It could be removed

File: splunklogger.js [hunk 4]
```diff
@@ -169,16 +250,66 @@ SplunkLogger.prototype._initializeConfig = function(config) {
...
+        ret.batchInterval = config.batchInterval || ret.batchInterval || defaultConfig.batchInterval;
+        ret.batchInterval = parseInt(ret.batchInterval, 10);
+        if (isNaN(ret.batchInterval)) {
+            throw new Error("Batch interval must be a number, found: " + ret.batchInterval);
+        }
+        else if (ret.batchInterval < 0) {
+            throw new Error("Batch interval must be a positive number, found: " + ret.batchInterval);
+        }
+
+        var startTimer = !this._timerID && ret.autoFlush && ret.batchInterval > 0;
```
Comment (itay) [inline splunklogger.js]: These are a bit too "cute" - you at least need to add a comment explaining the logic.

File: splunklogger.js [hunk 5]
```diff
@@ -451,17 +629,20 @@ SplunkLogger.prototype._sendEvents = function(context, callback) {
...
-    if (context.config.autoFlush) {
-        this.flush(callback);
-    }
-    else {
-        callback();
+    var batchOverSize = this.calculateBatchSize() > context.config.maxBatchSize;
+    var batchOverCount = this.contextQueue.length >= context.config.maxBatchCount && context.config.maxBatchCount > 0;
+
+    // Only flush if not using manual batching, and if the contextQueue is too large or has many events
+    if (context.config.autoFlush && !this._timerID && (batchOverSize || batchOverCount)) {
```
Comment (itay) [inline splunklogger.js]: I'm not sure this is right. Namely, if the batch is too big (by count or size), then even if there is a timer, you should still send it. You can use all three "batch" metrics at the same time.

File: splunklogger.js [hunk 6]
```diff
@@ -451,17 +629,20 @@ SplunkLogger.prototype._sendEvents = function(context, callback) {
...
+    // Store the context, and its estimated length
     this.contextQueue.push(context);
+    this.eventSizes.push(Buffer.byteLength(JSON.stringify(this._makeBody(context)), "utf8"));
 
-    if (context.config.autoFlush) {
-        this.flush(callback);
-    }
-    else {
-        callback();
+    var batchOverSize = this.calculateBatchSize() > context.config.maxBatchSize;
```
Comment (itay) [inline splunklogger.js]: I'm not sure I see why you need to calculate this every time. Namely, why do you need to keep a list of the batch sizes and not just keep the total batch size up until now?

File: splunklogger.js (hunk 5 above)
Comment (itay) [inline splunklogger.js]: Does this mean that `autoFlush` overrides any batch parameters (i.e. if autoFlush is false, any batch settings are ignored)?

File: splunklogger.js [hunk 7]
```diff
@@ -474,27 +655,32 @@ SplunkLogger.prototype.send = function (context, callback) {
...
 
-    // Use the batching setting from this.config
-    if (this.config.autoFlush) {
-        // TODO: handle case of multiple events with autoFlush off, flushing fast
-        // Just take the oldest event in the queue
-        context = this.contextQueue.pop();
-    }
-    else {
-        // Empty the event queue
+    var batchOverSize = this.config.maxBatchSize > 0 && this.calculateBatchSize() > this.config.maxBatchSize;
```
Comment (itay) [inline splunklogger.js]: It's unfortunate we have this logic duplicated.

File: splunklogger.js [hunk 8]
```diff
@@ -474,27 +655,32 @@ SplunkLogger.prototype.send = function (context, callback) {
...
+        this.eventSizes = [];
         var data = "";
         for (var i = 0; i < queue.length; i++) {
             data += JSON.stringify(this._makeBody(queue[i]));
         }
         context.message = data;
     }
+    // Just take the oldest event in the queue
+    else {
+        // TODO: handle case of multiple events with autoFlush off, flushing fast
```
Comment (itay) [inline splunklogger.js]: I don't understand this case. If I have autoFlush off, and I manually call `flush()`, it should just send everything thats queued up at the moment.

File: splunklogger.js [hunk 9]
```diff
@@ -513,17 +699,35 @@ SplunkLogger.prototype.flush = function (callback) {
...
         // point to the same memory block.
-        // The passedContext parameter could be named context to
-        // do this automatically, but the || notation adds a bit of clarity.
         if (err) {
-            that.error(err, passedContext || context);
+            that.error(err, context);
         }
         else {
+            // Validate the context again, right before using it
+            context = that._initializeContext(context);
```
Comment (itay) [inline splunklogger.js]: You say `validate` but yet it is called `initialize`

File: splunklogger.js [hunk 10]
```diff
@@ -513,17 +699,35 @@ SplunkLogger.prototype.flush = function (callback) {
...
+            context.requestOptions.headers["Authorization"] = "Splunk " + context.config.token;
```
Comment (itay) [inline splunklogger.js]: It's super weird you need to do this.

File: splunklogger.js [hunk 11]
```diff
@@ -513,17 +699,35 @@ SplunkLogger.prototype.flush = function (callback) {
...
+
+            if (isBatched) {
+                // Don't run _makeBody since we've already done that for batching
+                // Manually set the content-type header for batched requests, default is application/json
+                // since json is set to true.
+                context.requestOptions.headers["content-type"] = "application/x-www-form-urlencoded";
```
Comment (itay) [inline splunklogger.js]: This doesn't seem like the right header - should be JSON.

File: splunklogger.js [hunk 12]
```diff
@@ -513,17 +699,35 @@ SplunkLogger.prototype.flush = function (callback) {
...
         if (err) {
-            that.error(err, passedContext || context);
+            that.error(err, context);
         }
         else {
+            // Validate the context again, right before using it
+            context = that._initializeContext(context);
+            context.requestOptions.headers["Authorization"] = "Splunk " + context.config.token;
+
+            if (isBatched) {
```
Comment (itay) [inline splunklogger.js]: I also don't get this case here. Why do you distinguish between batch and unbatched?

File: test/test_config.js [hunk 13]
```diff
@@ -159,6 +163,177 @@ describe("SplunkLogger", function() {
             assert.strictEqual("important", logger.config.level);
             assert.strictEqual(false, logger.config.autoFlush);
             assert.strictEqual(8088, logger.config.port);
+            assert.strictEqual(0, logger.config.maxRetries);
+        });
+        it("should error when _enableInterval(NaN)", function() {
```
Comment (itay) [inline test/test_config.js]: Rename the test to match the function.

File: splunklogger.js (hunk 4 above)
Comment (shakeelmohamed) [inline splunklogger.js]: comments added

File: test/test_config.js [hunk 14]
```diff
@@ -159,6 +163,177 @@ describe("SplunkLogger", function() {
...
+            assert.strictEqual("localhost", logger.config.host);
+            assert.strictEqual("/services/collector/event/1.0", logger.config.path);
+            assert.strictEqual("https", logger.config.protocol);
+            assert.strictEqual("info", logger.config.level);
+            assert.strictEqual(true, logger.config.autoFlush);
+            assert.strictEqual(0, logger.config.batchInterval);
+            assert.strictEqual(8088, logger.config.port);
+            assert.strictEqual(0, logger.config.maxRetries);
+        });
+        it("should not set a batch interval timer with autoFlush off, & batchInterval set", function() {
```
Comment (itay) [inline test/test_config.js]: I wonder if this case (auto flush off with any kind of batching) should throw an error.

File: splunklogger.js (hunk 12 above)
Comment (shakeelmohamed) [inline splunklogger.js]: For isBatched we empty the entire queue, otherwise 1 event per request

Comment (itay) [inline splunklogger.js]: I understand what you're doing, I just don't understand why you're doing it. Why are you making a distinction here?

File: splunklogger.js (hunk 11 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Without this line, batched requests will get an error from Splunk saying "Invalid data format" with code 6.

Comment (itay) [inline splunklogger.js]: Huh? Why?

File: examples/retry.js [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+    },
+    // Severity is also optional
+    severity: "info"
+};
+
+console.log("Sending payload", payload);
+Logger.send(payload, function(err, resp, body) {
+    // If successful, body will be { text: 'Success', code: 0 }
+    console.log("Response from Splunk", body);
+});
```
Comment (geloizi) [inline examples/retry.js]: nit - there is no \n at eof

File: examples/retry.js [hunk 16]
```diff
@@ -0,0 +1,69 @@
+/*
```
Comment (geloizi) [inline examples/retry.js]: probably not for now, but doxygen has \copyright  tag. We may consider using it

File: examples/retry.js [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * This example shows how to configure retries with SplunkLogger.
+ */
+
+// Change to require("splunk-logging").Logger;
```
Comment (geloizi) [inline examples/retry.js]: have you forgotten to do this change?

File: examples/retry.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    level: "info",
+    maxRetries: 5
+};
+
+// Create a new logger
+var Logger = new SplunkLogger(config);
+
+Logger.error = function(err, context) {
+    // Handle errors here
+    console.log("error", err, "context", context);
```
Comment (geloizi) [inline examples/retry.js]: there is "console.error"

File: splunklogger.js [hunk 19]
```diff
@@ -42,7 +42,7 @@ function _err(err, context) {
  *     token: "your-token-here",
  *     name: "my application",
  *     host: "splunk.local",
- *     autoFlush: false
+ *     autoFlush: true
```
Comment (geloizi) [inline splunklogger.js]: curious - why this change?

File: splunklogger.js [hunk 20]
```diff
@@ -50,6 +50,7 @@ function _err(err, context) {
  * @property {object} config - Configuration settings for this <code>SplunkLogger</code> instance.
  * @property {function[]} middlewares - Middleware functions to run before sending data to Splunk.
  * @property {object[]} contextQueue - Queue of <code>context</code> objects to be sent to Splunk.
+ * @property {number[]} eventSizes - Parallel queue of the sizes of events stored in <code>contextQueue</code>.
```
Comment (geloizi) [inline splunklogger.js]: I'm not sure what is parallel queue... I saw parallel queues in Costco. Is that what you mean? :)

File: splunklogger.js [hunk 21]
```diff
@@ -107,6 +136,58 @@ var defaultRequestOptions = {
...
+/**
+ * Configures an interval timer to flush any events in
+ * <code>this.contextQueue</code> at the specified interval.
+ *
+ * param {Number} interval - The batch interval.
+ * @private
+ */
+SplunkLogger.prototype._enableTimer = function(interval) {
+    // Only enable the timer if possible
+    if (typeof interval === "number") {
```
Comment (geloizi) [inline splunklogger.js]: Up to you, but I'd make a set of validation functions for numbers, positive numbers, etc.

File: splunklogger.js [hunk 22]
```diff
@@ -107,6 +136,58 @@ var defaultRequestOptions = {
...
+            if (this._timerID) {
+                this._disableTimer();
+            }
+            
+            // If batch interval is changed, update the config property
+            if (this.config) {
+                this.config.batchInterval = interval;
+            }
+
+            this._timerDuration = interval;
```
Comment (geloizi) [inline splunklogger.js]: _timerInterval looks better to me. Feel free to ignore this comment if you still feel that "duration" is prefered (I'm absolutely sure your English is much-much better than mine so that's up to you).

File: splunklogger.js [hunk 23]
```diff
@@ -107,6 +136,58 @@ var defaultRequestOptions = {
...
+        this._timerDuration = 0;
+        this._timerID = null;
+    }
+};
+
+/**
+ * Configures an interval timer to flush any events in
+ * <code>this.contextQueue</code> at the specified interval.
+ *
+ * param {Number} interval - The batch interval.
```
Comment (geloizi) [inline splunklogger.js]: what measurement units do you use here? Milliseconds?

File: splunklogger.js [hunk 24]
```diff
@@ -169,16 +250,68 @@ SplunkLogger.prototype._initializeConfig = function(config) {
...
+        if (isNaN(ret.maxBatchCount)) {
+            throw new Error("Max batch count must be a number, found: " + ret.maxBatchCount);
+        }
+        else if (ret.maxBatchCount < 0) {
+            throw new Error("Max batch count must be a positive number, found: " + ret.maxBatchCount);
+        }
+
+        ret.maxBatchSize = config.maxBatchSize || ret.maxBatchSize || defaultConfig.maxBatchSize;
+        ret.maxBatchSize = parseInt(ret.maxBatchSize, 10);
+        if (isNaN(ret.maxBatchSize)) {
```
Comment (geloizi) [inline splunklogger.js]: what if maxBatchSize = 1K and maxBatchCount isn't specified? 
In .NET/Java we set maxBatchCount to MaxInt in this case.
Same for the interval

Comment (geloizi) [inline splunklogger.js]: Honestly, I'd make several validation functions.

File: splunklogger.js [hunk 25]
```diff
@@ -360,8 +493,34 @@ SplunkLogger.prototype.use = function(middleware) {
...
 /**
+ * Estimates the size in bytes of the events in the
+ * <code>queue</code> parameter as if they were sent
+ * in a single HTTP request.
+ *
+ * @returns {number} the estimated size in bytes.
+ */
+SplunkLogger.prototype.calculateBatchSize = function() {
+    var size = 0;
+    for (var i = 0; i < this.eventSizes.length; i++) {
```
Comment (geloizi) [inline splunklogger.js]: nit, you could use Array.reduce() here

File: splunklogger.js [hunk 26]
```diff
@@ -369,34 +528,55 @@ SplunkLogger.prototype.use = function(middleware) {
...
+                _response = resp;
+                _body = body;
+
+                // Try to parse an error response from Splunk
+                if (!requestError && body && body.code.toString() !== "0") {
+                    splunkError = new Error(body.text);
+                    splunkError.code = body.code;
+                }
+
+                // Request error (non-200), retry if numRetries hasn't exceeded the limit
```
Comment (geloizi) [inline splunklogger.js]: There is a subtle problem here. There is no point to retry when the server returns http input error. For example if parser fails it will fail in each retry attempt.
.NET/Java have "error whitelist" logic. It's really simple ...

File: splunklogger.js (hunk 8 above)
Comment (geloizi) [inline splunklogger.js]: I agree with @itay . This condition looks strange.

File: utils.js [hunk 27]
```diff
@@ -128,4 +128,83 @@ utils.chain = function(tasks, callback) {
...
+ * @param {object} [self] - An object to bind the <code>fn</code> function parameter to.
+ * @param {object} [fn] - A function to bind to the <code>self</code> argument.
+ */
+utils.bind = function(self, fn) {
+    return function () {
+        return fn.apply(self, arguments);
+    };
+};
+
 module.exports = utils;
```
Comment (geloizi) [inline utils.js]: nit: no \n at eof

File: splunklogger.js (hunk 19 above)
Comment (shakeelmohamed) [inline splunklogger.js]: `true` is the correct default, this is part of a doc comment

File: splunklogger.js (hunk 20 above)
Comment (shakeelmohamed) [inline splunklogger.js]: :laughing: This will go away with some refactoring

File: splunklogger.js (hunk 25 above)
Comment (shakeelmohamed) [inline splunklogger.js]: This will go away with some refactoring, I will use a running count like Java/.Net

File: splunklogger.js (hunk 26 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Any errors from Event Collector will still be 200 responses, so requestError will be falsey. The if condition will be false in that case. I will make the code/comment more clear

File: splunklogger.js (hunk 10 above)
Comment (shakeelmohamed) [inline splunklogger.js]: We have to add the auth header somewhere... not sure what's weird about this

File: splunklogger.js (hunk 11 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Not sure why the JSON content-type doesn't work, but that header is force by the option `json=true`. Another way to do this would be committing the `json=true` request option and manually calling `JSON.parse()` on every response.

File: examples/retry.js (hunk 15 above)
Comment (shakeelmohamed) [inline examples/retry.js]: Our other logging libs + SDKs don't always have a blank line at the end of the file, not going to worry about it

File: splunklogger.js (hunk 2 above)
Comment (shakeelmohamed) [inline splunklogger.js]: I tried this and ran into strange scope issues, punting for now.

File: splunklogger.js (hunk 3 above)
Comment (shakeelmohamed) [inline splunklogger.js]: done

File: splunklogger.js (hunk 5 above)
Comment (shakeelmohamed) [inline splunklogger.js]: fixed

File: splunklogger.js (hunk 6 above)
Comment (shakeelmohamed) [inline splunklogger.js]: fixed

File: splunklogger.js (hunk 7 above)
Comment (shakeelmohamed) [inline splunklogger.js]: It's no longer duplicated :+1:

File: splunklogger.js (hunk 8 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Done, now treating everything like a batch in `flush()`

File: splunklogger.js (hunk 9 above)
Comment (shakeelmohamed) [inline splunklogger.js]: fixed

File: splunklogger.js (hunk 12 above)
Comment (shakeelmohamed) [inline splunklogger.js]: fixed

File: test/test_config.js (hunk 13 above)
Comment (shakeelmohamed) [inline test/test_config.js]: done

File: test/test_config.js (hunk 14 above)
Comment (shakeelmohamed) [inline test/test_config.js]: done

File: examples/retry.js (hunk 17 above)
Comment (shakeelmohamed) [inline examples/retry.js]: Nope, this is by design

File: splunklogger.js (hunk 20 above)
Comment (shakeelmohamed) [inline splunklogger.js]: all gone now

File: splunklogger.js (hunk 21 above)
Comment (shakeelmohamed) [inline splunklogger.js]: done

File: splunklogger.js (hunk 23 above)
Comment (shakeelmohamed) [inline splunklogger.js]: yes & done

File: splunklogger.js (hunk 25 above)
Comment (shakeelmohamed) [inline splunklogger.js]: all gone now

File: splunklogger.js (hunk 26 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Done

File: utils.js (hunk 27 above)
Comment (shakeelmohamed) [inline utils.js]: Rejected, see above :smile:

File: splunklogger.js (hunk 24 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Added lots of validation util functions.

I took a slightly different approach for that scenario by offering recommended defaults. Whereas in Java & .Net we define public const/static values for recommended batching, but they're never used in the constructor.

File: examples/basic.js [hunk 28]
```diff
@@ -23,20 +23,11 @@ var SplunkLogger = require("../index").Logger;
...
 var config = {
     token: "your-token-here",
-    host: "localhost",
-    path: "/services/collector/event/1.0",
-    protocol: "https",
-    port: 8088,
-    level: "info",
-    autoFlush: true
+    url: "https://localhost:8088",
+    maxBatchCount: 1 // Send events 1 at a time
```
Comment (itay) [inline examples/basic.js]: Why do you need to set this?

File: examples/basic.js [hunk 29]
```diff
@@ -66,6 +57,34 @@ var payload = {
...
+ * will immediately send the payload.
+ * 
+ * The underlying HTTP POST request is made to
+ *
+ *     https://localhost:8088/services/collector/event/1.0
+ *
+ * with the following body
+ *
+ *     {
+ *         "metadata": {
```
Comment (itay) [inline examples/basic.js]: This definitely does not look right. There is no `metadata` object in the HTTP Input - those keys should be in the top level object, adjacent to `event`.

File: examples/all_batching.js [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+    },
+    metadata: payload.metadata
+};
+
+console.log("Queuing second payload", payload2);
+// Don't need a callback here
+Logger.send(payload2);
+
+/**
+ * Since we've configured autoFlush, we don't need
```
Comment (itay) [inline examples/all_batching.js]: Here you refer to `autoFlush`, but do you mean batching?

File: examples/custom_format.js [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+/**
+ * Override the default eventFormatter() function,
+ * which takes a message and severity, returning
+ * any type - string or object are recommended.
+ *
+ * The message parameter can be any type. It will
+ * be whatever was passed to Logger.send().
+ * Severity will always be a string.
+ *
+ * In this example, we're
```
Comment (itay) [inline examples/custom_format.js]: ....

File: examples/retry.js [hunk 32]
```diff
@@ -75,7 +56,7 @@ var payload = {
         source: "chicken coop",
         sourcetype: "httpevent",
         index: "main",
-        host: "farm.local"
+        host: "farm.local",
```
Comment (itay) [inline examples/retry.js]: This is a bug.

File: examples/basic.js (hunk 28 above)
Comment (shakeelmohamed) [inline examples/basic.js]: If none of the batch settings are configured, and `autoFlush === true` (the default) it's basically not manual batching

File: splunklogger.js [hunk 33]
```diff
@@ -162,35 +249,57 @@ SplunkLogger.prototype._initializeConfig = function(config) {
...
-            ret.autoFlush = config.autoFlush;
+        ret.token = utils.orByProp("token", config, ret);
+        ret.name = utils.orByProp("name", config, ret, defaultConfig);
+        ret.level = utils.orByProp("level", config, ret, defaultConfig);
+
+        ret.host = utils.orByProp("host", config, ret, defaultConfig);
+        ret.path = utils.orByProp("path", config, ret, defaultConfig);
+        ret.protocol = utils.orByProp("protocol", config, ret, defaultConfig);
+        ret.port = utils.orByProp("port", config, ret, defaultConfig);
+        ret.port = utils.validateNonNegativeInt(ret.port, "Port");
```
Comment (itay) [inline splunklogger.js]: is the capital `P` on purpose?

File: examples/basic.js (hunk 29 above)
Comment (shakeelmohamed) [inline examples/basic.js]: Wow good catch, fixed.

File: examples/all_batching.js (hunk 30 above)
Comment (shakeelmohamed) [inline examples/all_batching.js]: fixed

File: splunklogger.js [hunk 34]
```diff
@@ -451,82 +556,43 @@ SplunkLogger.prototype._sendEvents = function(context, callback) {
  * @throws Will throw an error if the <code>context</code> parameter is malformed.
  * @public
  */
-SplunkLogger.prototype.send = function (context, callback) {
-    callback = callback || function(){};
+SplunkLogger.prototype.send = function(context, callback) {
     context = this._initializeContext(context);
     
+    // Store the context, and its estimated length
     this.contextQueue.push(context);
```
Comment (itay) [inline splunklogger.js]: Why do you need `contextQueue`?

File: splunklogger.js [hunk 35]
```diff
@@ -451,82 +556,43 @@ SplunkLogger.prototype._sendEvents = function(context, callback) {
...
-    var callbacks = [];
-    for (var j = 0; j < this.middlewares.length; j++) {
-        callbacks[j] = this.middlewares[j];
+    // Empty the queue, reset the eventsBatchSize
+    var queue = this.contextQueue;
+    this.contextQueue = [];
+    this.eventsBatchSize = 0;
+    var data = "";
+    for (var i = 0; i < queue.length; i++) {
+        data += JSON.stringify(this._makeBody(queue[i]));
```
Comment (itay) [inline splunklogger.js]: You've already stringified it once when you counted the bytes - why are you doing it again?

File: splunklogger.js (hunk 33 above)
Comment (shakeelmohamed) [inline splunklogger.js]: Yes, it actually doesn't matter since the string is only used for error messages.

File: splunklogger.js (hunk 35 above)
Comment (shakeelmohamed) [inline splunklogger.js]: refactored, done

File: splunklogger.js (hunk 34 above)
Comment (shakeelmohamed) [inline splunklogger.js]: simplified, done

File: examples/retry.js (hunk 32 above)
Comment (shakeelmohamed) [inline examples/retry.js]: Fixed

**Comment by shakeelmohamed** [comment]
Closing this PR, opening another one with updated readme, examples and version numbers.
